apps/pipelines/nodes/history_middleware.py

```diff
@@ -0,0 +1,121 @@
+from __future__ import annotations
+
+import logging
+from typing import TYPE_CHECKING
+
+from langchain.agents.middleware.summarization import SummarizationMiddleware
+from langchain_core.messages import BaseMessage, HumanMessage, RemoveMessage
+from langgraph.graph.message import (
+    REMOVE_ALL_MESSAGES,
+)
+
+from apps.chat.conversation import COMPRESSION_MARKER
+
+if TYPE_CHECKING:
+    from apps.pipelines.nodes.nodes import PipelineNode
+
+logger = logging.getLogger("ocs.bots.summarization")
+
+
+class BaseNodeHistoryMiddleware(SummarizationMiddleware):
+    """
+    Middleware to summarize chat history based on node configuration.
+
+    History is loaded into the state in the before_agent step, and summaries are persisted
+    to the database in the before_model step, if a summary is generated.
+    """
+
+    def __init__(self, session, node: PipelineNode, **kwargs):
+        super().__init__(model=node.get_chat_model(), **kwargs)
+        self.session = session
+        self.node = node
+
+    def before_agent(self, state, runtime):
+        return {
+            "messages": [
+                # Since this response will get merged with the existing state messages, we cannot simply append the
+                # history to the user's message. We need to replace the full message history in the state.
+                # See https://github.com/langchain-ai/langchain/blob/c63f23d2339b2604edc9ae1d9f7faf7d6cc7dc78/libs/langchain_v1/langchain/agents/middleware/summarization.py#L286-L292
+                RemoveMessage(id=REMOVE_ALL_MESSAGES),
+                *self.node.get_history(self.session),
+                *state["messages"],
+            ]
+        }
+
+    def before_model(self, state, runtime):
+        result = super().before_model(state, runtime)
+        if result is not None:
+            # A result means that a summary was created
+            self.persist_summary(result["messages"])
+
+        return result
+
+    def persist_summary(self, messages: list[BaseMessage]):
+        checkpoint_message_id = self._find_latest_message_db_id(messages)
+        if not checkpoint_message_id:
+            logger.exception(
```

Comment:
Using `logger.exception` here isn't great because it will use `sys.exc_info` to get the exception info which can be misleading when there isn't actually an exception (I speak from experience).

If you want to log the trace you can use `logger.error("", stack_info=True)`.

See https://docs.python.org/3/library/logging.html#logging.Logger.debug